### PR TITLE
treewide: enable strictDeps in bootstrap packages

### PR DESCRIPTION
**Pull request by Artturin**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 173909` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hello</li>
  </ul>
</details>

**Comment by Artturin** [comment]
alright, only one more
```
$ nix-diff $(nix eval --raw ".#bash.drvPath") $(nix eval --raw --expr 'with import ./. { config = { strictDepsByDefault = true; }; }; bash.drvPath' --impure)
- /nix/store/1s8vhyfq82127wrkrzqnl4jrn68kc74a-bash-5.1-p16.drv:{out}
+ /nix/store/mla63jkfy61507fvbxa2x92zskxrkpsn-bash-5.1-p16.drv:{out}
• The input derivation named `bootstrap-stage4-gcc-wrapper-11.3.0` differs
  - /nix/store/npj5218170k7nqr744ji146r71cnfhnc-bootstrap-stage4-gcc-wrapper-11.3.0.drv:{out}
  + /nix/store/19l65594r0r68dyyaikpr2ckbxgw4ck9-bootstrap-stage4-gcc-wrapper-11.3.0.drv:{out}
  • The input derivation named `gcc-11.3.0` differs
    - /nix/store/6hl94g48qsxbv2w0h4v6gdxn1g1s9hsi-gcc-11.3.0.drv:{info,lib,man,out}
    + /nix/store/6j2fmb1cj46bnivyskd64dlf2xbgz8fm-gcc-11.3.0.drv:{info,lib,man,out}
    • The environments do not match:
        strictDeps=1
  • Skipping environment comparison
• The input derivation named `bootstrap-stage4-stdenv-linux` differs
  - /nix/store/2rsa0xsw9k127hq276sqy74gzpc0wcbw-bootstrap-stage4-stdenv-linux.drv:{out}
  + /nix/store/y5bcnivsn21qxi8hp9rh0r7mzzr5dgp4-bootstrap-stage4-stdenv-linux.drv:{out}
  • The input derivation named `bootstrap-stage4-gcc-wrapper-11.3.0` differs
    • These two derivations have already been compared
  • Skipping environment comparison
• Skipping environment comparison
```

i'll probably do gcc in a separate PR

**Comment by Artturin** [comment]
@ofborg build bash

File: pkgs/development/tools/parsing/bison/default.nix [hunk 1]
```diff
@@ -18,6 +18,11 @@ stdenv.mkDerivation rec {
   # Otherwise tests fail due to incorrect unicode symbol oconversion.
   configurePlatforms = [ "build" "host" ];
 
+  # there's a /bin/sh shebang in bin/yacc which when no strictDeps is patched with the build stdenv shell
+  # however when cross-compiling it would still be patched with the build stdenv shell which would be wrong
+  # cannot add bash to buildInputs due to infinite recursion
+  strictDeps = stdenv.hostPlatform != stdenv.buildPlatform;
```
Comment (Ericson2314) [inline pkgs/development/tools/parsing/bison/default.nix:24]: I would do
```nix
buildInputs = [
  # Hack, cannot depend on bash in the native build without
  # causing infinite recursion during bootstrapping
  (if stdenv.hostPlatform != stdenv.buildPlatform then bash else stdenv.shell)
];
```
instead

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 2]
```diff
@@ -104,8 +104,6 @@ let
 
   version = with sourceVersion; "${major}.${minor}.${patch}${suffix}";
 
-  strictDeps = true;
-
```
Comment (Ericson2314) [inline pkgs/development/interpreters/python/cpython/default.nix:108]: How come?

File: pkgs/development/interpreters/python/default.nix [hunk 3]
```diff
@@ -231,6 +231,8 @@ in {
     enableLTO = false;
     mimetypesSupport = false;
   } // sources.python39)).overrideAttrs(old: {
+    # TODO(@Artturin): Add this to the main cpython expr
+    strictDeps = true;
```
Comment (Ericson2314) [inline pkgs/development/interpreters/python/default.nix:235]: Wasn't it there before?

File: pkgs/development/interpreters/python/cpython/default.nix (hunk 2 above)
Comment (Artturin) [reply pkgs/development/interpreters/python/cpython/default.nix:108]: see https://github.com/NixOS/nixpkgs/commit/73dee3ad

File: pkgs/development/tools/parsing/bison/default.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/development/tools/parsing/bison/default.nix:24]:
```
$ nix build ".#bison"
bison> /nix/store/nmca9y3m39hypn342bwzwm6c4yd5y9lp-bootstrap-stage1-stdenv-linux/setup: line 528: source: /nix/store/p4s4jf7aq6v6z9iazll1aiqwb34aqxq9-bootstrap-tools/bin/bash: cannot execute binary file
note: keeping build directory '/comp-temp/nix-build-bison-3.8.2.drv-4'
error: builder for '/nix/store/xgmr7z154q3fmbkyr97m0na9jcyjvixf-bison-3.8.2.drv' failed with exit code 126;
       last 1 log lines:
       > /nix/store/nmca9y3m39hypn342bwzwm6c4yd5y9lp-bootstrap-stage1-stdenv-linux/setup: line 528: source: /nix/store/p4s4jf7aq6v6z9iazll1aiqwb34aqxq9-bootstrap-tools/bin/bash: cannot execute binary file
       For full logs, run 'nix log /nix/store/xgmr7z154q3fmbkyr97m0na9jcyjvixf-bison-3.8.2.drv'.
```

https://github.com/NixOS/nixpkgs/blob/7283d3942f7947949a131cfa81e5fd0688af3de3/pkgs/stdenv/generic/setup.sh#L515

Comment (Artturin) [reply pkgs/development/tools/parsing/bison/default.nix:24]:
```
nix-repl> stdenv.shell
"/nix/store/wpcyy8zajlr47y0vmljwj3ppgck0rb7k-bash-5.1-p16/bin/bash"

nix-repl> stdenv.shellPackage
«derivation /nix/store/5qnnrqd025j0swnvhiqxg7sda2x9n9yl-bash-5.1-p16.drv»
```

changing it to `stdenv.shellPackage` gives

```
$ nix build ".#bison"
error: attribute 'shellPackage' missing

       at /nix/store/czfggskqip31vji4mwgpvgp0474j79yl-source/pkgs/development/tools/parsing/bison/default.nix:27:68:

           26|     # causing infinite recursion during bootstrapping
           27|     (if stdenv.hostPlatform != stdenv.buildPlatform then bash else stdenv.shellPackage)
             |                                                                    ^
           28|   ];
(use '--show-trace' to show detailed location information)
```

File: pkgs/development/interpreters/python/cpython/default.nix (hunk 2 above)
Comment (Ericson2314) [reply pkgs/development/interpreters/python/cpython/default.nix:108]: OK

**Review by trofi — APPROVED** [review]
Built ~2k packages without any issues. LGTM.

File: pkgs/development/tools/parsing/bison/default.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/development/tools/parsing/bison/default.nix:24]: lets leave it as is for now and then figure it out in a separate pr

**Comment by Artturin** [comment]
@ofborg build bash

File: pkgs/development/tools/parsing/bison/default.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/development/tools/parsing/bison/default.nix:24]: Sounds good.

**Comment by Artturin** [comment]
@ofborg eval

the failure is unrelated

```
Test Summary Report-------------------
t2200-add-update.sh                              (Wstat: 256 Tests: 17 Failed: 1)
  Failed test:  15  Non-zero exit status: 1
Files=949, Tests=24140, 57 wallclock secs ( 9.29 usr  3.23 sys + 544.43 cusr 1176.85 csys = 1733.80 CPU)Result: FAILmake[1]: *** [Makefile:56: prove] Error 1make[1]: Leaving directory '/build/git-2.36.0/t'make: *** [Makefile:2949: test] Error 2error: builder for '/nix/store/ay96gmi1fn62vqjhhcr2vkpkya74k3h7-git-minimal-2.36.0.drv' failed with exit code 2;       last 10 log lines:       > Test Summary Report       > -------------------       > t2200-add-update.sh                              (Wstat: 256 Tests: 17 Failed: 1)       >   Failed test:  15       >   Non-zero exit status: 1       > Files=949, Tests=24140, 57 wallclock secs ( 9.29 usr  3.23 sys + 544.43 cusr 1176.85 csys = 1733.80 CPU)       > Result: FAIL       > make[1]: *** [Makefile:56: prove] Error 1       > make[1]: Leaving directory '/build/git-2.36.0/t'       > make: *** [Makefile:2949: test] Error 2       For full logs, run 'nix log /nix/store/ay96gmi1fn62vqjhhcr2vkpkya74k3h7-git-minimal-2.36.0.drv'.
```

bash was built succesfully

File: pkgs/development/tools/parsing/bison/default.nix (hunk 1 above)
Comment (Mindavi) [reply pkgs/development/tools/parsing/bison/default.nix:24]: This is already true by default, but making it explicit sounds ok to me

**Review by Mindavi — APPROVED** [review]
Happy to see this work being done ✨

File: pkgs/development/tools/parsing/bison/default.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/development/tools/parsing/bison/default.nix:24]: making it explicit prevents a rebuild when `strictDepsByDefault` is true

**Comment by Artturin** [comment]
2 approvals
bash and sway compile successfully
bash cross-compiles
output diffs look ok (checked with [diffing](https://github.com/Artturin/diffing))

therefore i will soon merge 

diffs checked by doing  `git diff --name-only HEAD HEAD~6 | rev | cut -d '/' -f 2 | rev > attrlist` then cleaned it up manually
and then ran `diffing --force --file attrlist`

```
nukeReferences
pkg-config
perl
python
expat
glibc
gmp
isl
libelf
libffi
libidn2
libmpc
libunistring
mpfr
zlib
autoconf-archive
autoconf
automake
binutils
gnum4
libtool
patchelf
pkg-config
texinfo
bison
linuxHeaders
bzip2
xz
file
which
```

**Review by Ericson2314 — APPROVED** [review]

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/185972

**Pull request merged**